crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs

```diff
@@ -0,0 +1,882 @@
+                element_range
+            }
+        };
+        Self {
+            value,
+            preceding_comment_ranges,
+            element_range,
+            total_range,
+            end_of_line_comments,
+        }
+    }
+
+    fn with_no_comments(value: String, element_range: TextRange) -> Self {
+        Self::new(value, vec![], element_range, None)
+    }
+}
+
+impl Ranged for DunderAllItem {
+    fn range(&self) -> TextRange {
+        self.total_range
+    }
+}
+
+fn multiline_dunder_all_prelude(
+    first_item_start_offset: TextSize,
+    newline: &str,
+    dunder_all_offset: TextSize,
+    locator: &Locator,
+) -> String {
+    let prelude_end = {
+        let first_item_line_offset = locator.line_start(first_item_start_offset);
+        if first_item_line_offset == locator.line_start(dunder_all_offset) {
+            first_item_start_offset
+        } else {
+            first_item_line_offset
+        }
+    };
+    let prelude = locator.slice(TextRange::new(dunder_all_offset, prelude_end));
+    format!("{}{}", prelude.trim_end(), newline)
+}
+
+fn join_multiline_dunder_all_items(
+    sorted_items: &[DunderAllItem],
+    locator: &Locator,
+    item_indent: &str,
+    newline: &str,
+    needs_trailing_comma: bool,
+) -> String {
+    let max_index = sorted_items.len() - 1;
+
+    let mut new_dunder_all = String::new();
+    for (i, item) in sorted_items.iter().enumerate() {
+        let is_final_item = i == max_index;
+        for comment_range in &item.preceding_comment_ranges {
+            new_dunder_all.push_str(item_indent);
+            new_dunder_all.push_str(locator.slice(comment_range));
+            new_dunder_all.push_str(newline);
+        }
+        new_dunder_all.push_str(item_indent);
+        new_dunder_all.push_str(locator.slice(item.element_range));
+        if !is_final_item || needs_trailing_comma {
+            new_dunder_all.push(',');
+        }
+        if let Some(trailing_comments) = item.end_of_line_comments {
+            new_dunder_all.push_str(locator.slice(trailing_comments));
+        }
+        if !is_final_item {
+            new_dunder_all.push_str(newline);
+        }
+    }
+    new_dunder_all
+}
+
+fn multiline_dunder_all_postlude<'a>(
+    last_item_end_offset: TextSize,
+    newline: &str,
+    leading_indent: &str,
+    item_indent: &str,
+    dunder_all_range_end: TextSize,
+    locator: &'a Locator,
+) -> Cow<'a, str> {
+    let postlude_start = {
+        let last_item_line_offset = locator.line_end(last_item_end_offset);
+        if last_item_line_offset == locator.line_end(dunder_all_range_end) {
+            last_item_end_offset
+        } else {
+            last_item_line_offset
+        }
+    };
+    let postlude = locator.slice(TextRange::new(postlude_start, dunder_all_range_end));
+    if !postlude.starts_with(newline) {
+        return Cow::Borrowed(postlude);
+    }
+    if TextSize::of(leading_indentation(postlude.trim_start_matches(newline)))
```

Comment:
The `newline` returned by `Stylist` is the newline used for fixing, but it may or may not be the newline used in the `postlude`. For example, `Stylist` detects `\n` for a file with mixed line breaks if the first line uses `\n` but `postlude` could use `\r\n`, in which case this path incorrectly assumes that the postlude doesn't start with a newline. 

I don't understnad the code well enough to understand if that's a problem. If it is, changing it to `!postlude.starts_with(['\n', '\r'])` might be enough.

Reply:
Thanks, good call!